### PR TITLE
Feature/buy asset

### PR DESCRIPTION
**Pull request by Lex-23**

### CODE REVIEW

File: cryptostock/utils/services.py [hunk 1]
```diff
@@ -104,3 +105,42 @@ def get_offers_with_related_items(request):
...
+
+
+def _update_broker_account_after_purchase(asset, broker, count, deal_total_price):
+    broker.cash_balance -= deal_total_price
+    try:
+        broker_wallet_record = broker.wallet.wallet_record.get(asset=asset)
+    except WalletRecord.DoesNotExist:
+        broker_wallet_record = WalletRecord.objects.create(
+            asset=asset, wallet=broker.wallet
+        )
```
Comment (Vladgasik) [inline cryptostock/utils/services.py:136]:
```suggestion
    broker_wallet_record, _= WalletRecord.objects.get_or_create(asset=asset, wallet=broker.wallet)
        
```

File: cryptostock/utils/services.py [hunk 2]
```diff
@@ -104,3 +105,42 @@ def get_offers_with_related_items(request):
...
+    broker_wallet_record.count += count
+    broker_wallet_record.save()
+    broker.save()
+
+
+def _purchase_asset_check(asset: dict):
+    try:
+        Asset.objects.get(name=asset["name"])
+    except Asset.DoesNotExist:
+        Asset.objects.create(name=asset["name"], description=asset["description"])
```
Comment (Vladgasik) [inline cryptostock/utils/services.py:146]: Use `get_or_create`

File: cryptostock/account/migrations/0005_purchasedashboard.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
+# Generated by Django 3.2.8 on 2021-12-09 14:23
```
Comment (vadim-denisov) [inline cryptostock/account/migrations/0005_purchasedashboard.py:1]: Should it be squashed?

File: cryptostock/market/models.py [hunk 4]
```diff
@@ -57,7 +59,14 @@ def get_asset(self, name):
...
     def buy(self, name, count):
-        return {"asset": self.get_asset(name), "count": count}
+        asset = self.get_asset(name)
+        total_price = (
+            decimal.Decimal(asset["price"]).quantize(
+                decimal.Decimal("0.01"), rounding=decimal.ROUND_UP
+            )
+            * count
+        )
+        return {"asset": asset, "count": count, "total_price": total_price}
```
Comment (vadim-denisov) [inline cryptostock/market/models.py:69]: Does it work?
```suggestion
        return BuyResponse(asset=asset, count=count, total_price=total_price)
```

File: cryptostock/market/tests/test_buy_asset_from_market.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
+from unittest.mock import MagicMock
+
+from market.models import YahooMarket
+from market.tests.factory import MarketFactory
+
+
+def test_buy_asset_from_market(auth_broker, broker_account, assets_list):
+    # TDD
```
Comment (vadim-denisov) [inline cryptostock/market/tests/test_buy_asset_from_market.py:8]:
```suggestion
```

File: cryptostock/utils/services.py (hunk 2 above)
Comment (Lex-23) [reply cryptostock/utils/services.py:146]: ok

File: cryptostock/market/tests/test_buy_asset_from_market.py (hunk 5 above)
Comment (Lex-23) [reply cryptostock/market/tests/test_buy_asset_from_market.py:8]: I wrote this permanently for understanding)

File: cryptostock/utils/services.py (hunk 1 above)
Comment (Lex-23) [reply cryptostock/utils/services.py:136]: Thanks for method)

File: cryptostock/account/migrations/0005_purchasedashboard.py (hunk 3 above)
Comment (Lex-23) [reply cryptostock/account/migrations/0005_purchasedashboard.py:1]: maybe)

File: cryptostock/market/models.py (hunk 4 above)
Comment (Lex-23) [reply cryptostock/market/models.py:69]: Yep

File: cryptostock/account/migrations/0005_purchasedashboard.py (hunk 3 above)
Comment (Lex-23) [reply cryptostock/account/migrations/0005_purchasedashboard.py:1]: Not(

**Comment by Lex-23** [comment]
[Write logic for BUY ASSETS from market](https://trello.com/c/nzNYUyZA/16-write-logic-for-buy-assets-from-market)

File: cryptostock/asset/models.py [hunk 6]
```diff
@@ -3,7 +3,10 @@
 
 class Asset(models.Model):
     name = models.CharField(max_length=50, unique=True)
-    description = models.TextField(blank=True, null=True)
+    description = models.CharField(max_length=250, blank=True, null=True)
+
+    class Meta:
+        unique_together = ("name", "description")
```
Comment (vadim-denisov) [inline cryptostock/asset/models.py:9]: name is unique. Why do we need this statement?

File: cryptostock/utils/services.py [hunk 7]
```diff
@@ -104,3 +106,37 @@ def get_offers_with_related_items(request):
...
+
+
+def _update_broker_account_after_purchase(asset, broker, count, deal_total_price):
+    broker_wallet_record, created = WalletRecord.objects.get_or_create(
+        asset=asset, wallet=broker.wallet
+    )
+    broker_wallet_record.count += count
+    broker_wallet_record.save()
+    broker.cash_balance -= deal_total_price
+    broker.save()
```
Comment (vadim-denisov) [inline cryptostock/utils/services.py:142]: Should be transaction

File: cryptostock/asset/models.py (hunk 6 above)
Comment (Lex-23) [reply cryptostock/asset/models.py:9]: maybe you`re right. It is reduntant

File: cryptostock/utils/services.py (hunk 7 above)
Comment (Lex-23) [reply cryptostock/utils/services.py:142]: ok

**Review by vadim-denisov — APPROVED** [review]

**Pull request merged**